classes/local/datasource/base.php

```diff
@@ -0,0 +1,230 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace mod_cms\local\datasource;
+
+use mod_cms\local\model\cms;
+use mod_cms\local\model\cms_types;
+
+/**
+ * Base class for data sources.
+ *
+ * @package   mod_cms
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2023, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+abstract class base {
+
+    /** Datasources provided by this plugin. */
+    public const BUILTIN_DATASOURCES = [
+        'site',
+        'customfield',
+        'image',
+    ];
+
+    /** @var array List of datasource class names in use. */
+    protected static $datasourceclasses = [];
+
+    /**
+     * Create the list of datasources to be used.
+     */
+    public static function register_datasources() {
+        // Only need to do this once.
+        if (!empty(self::$datasourceclasses)) {
+            return;
+        }
+
+        // Register the datasources that are defines in this plugin.
+        foreach (self::BUILTIN_DATASOURCES as $classname) {
+            self::add_datasource_class(__NAMESPACE__ . '\\' . $classname);
+        }
+
+        // Regsiter any datasources defined in other plugins.
+        $plugintypes = get_plugins_with_function('modcms_datasources', 'lib.php');
+        foreach ($plugintypes as $plugins) {
+            foreach ($plugins as $pluginfunction) {
+                $result = $pluginfunction();
+                foreach ($result as $classname) {
+                    self::add_datasource_class($classname);
+                }
+            }
+        }
+    }
+
+    /**
+     * Register a datasource.
+     *
+     * @param string $classname
+     * @throws \moodle_exception
+     */
+    public static function add_datasource_class(string $classname) {
+        // Test for existence.
+        if (!class_exists($classname)) {
+            throw new \moodle_exception('error:class_missing', 'mod_cms', '', $classname);
+        }
+        // Test for inheritance.
+        if (!is_subclass_of($classname, __NAMESPACE__ . '\\' . 'base')) {
```

Comment:
Pardon if my understanding is not correct here, but if we're checking for inheritance, does that mean any class that wants to register a datasource needs to implement mod_cms\local\datasource\base? Wouldn't that sort of defeat the purpose of "no plugin dependency"?

Reply:
A plug-in that registers a new source doesn’t need to declare a hard plugin dependancy. That file will only get auto loaded by mod_cms so if it’s not installed then nothing will hard break

Reply:
By hard dependencies I mean it is declared in version.php and it won’t install. By soft dependency I mean it can be installed and will gracefully not work but warn you in some nice way. We do not need a hard dependency

ovjectfs has a soft dependency on local_aws for example

Reply:
I can easily imagine a world where some local plug-in does a bunch of stuff and also makes available data to mod_cms if it is installed, but if you don’t have it installed and don’t need it then everything else you want still works just fine without it

Reply:
In my mind it just doesn’t work with the way I think callbacks vs sub plugins work. And again this is just how I see things, i.e. callback functions get a data structure (could be an object, array, etc) and then is able to manipulate it to achieve a desired effect. A sub plugin on the other hand provides additional functionality to the main plugin. In this case a sub plugin architecture seems more appropriate? I mean I get your “plugin does many things but also adds something else to mod_cms” but that just seems a bit weird to me

Reply:
It works just fine there is plenty of precedent and sub plugins are a pain to extend because then you have to have two plugins and you have to fork mod_cms and it’s all ugly. What we are setting up is the right way